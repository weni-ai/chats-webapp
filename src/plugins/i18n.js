import Vue from 'vue';
import VueI18n from 'vue-i18n';

// eslint-disable-next-line camelcase
import pt_br from '@/locales/pt_br.json';
import en from '@/locales/en.json';
import es from '@/locales/es.json';

import moment from 'moment';

import('moment/locale/pt-br');
import('moment/locale/es');

moment.locale('pt-br');

Vue.use(VueI18n);

const messages = {
  'pt-br': pt_br,
  en,
  es,
};

const i18n = new VueI18n({
  locale: 'pt-br',
  fallbackLocale: 'pt-br',
  messages,
  dateTimeFormats: {
    'pt-br': {
      short: {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
      },
      long: {
        month: 'long',
        day: '2-digit',
      },
    },
    en: {
      short: {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
      },
      long: {
        month: 'long',
        day: '2-digit',
      },
    },
    es: {
      short: {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
      },
      long: {
        month: 'long',
        day: '2-digit',
      },
    },
  },
});

export default i18n;
