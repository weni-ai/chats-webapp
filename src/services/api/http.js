import axios from 'axios';

const client = axios.create({
  baseURL: process.env.VUE_APP_CHATS_API_URL,
});

const token = process.env.VUE_APP_CHATS_API_TOKEN;
const isProduction = process.env.NODE_ENV === 'production';

if (token) client.defaults.headers.Authorization = `${isProduction ? 'Bearer' : 'Token'} ${token}`;

export default client;
