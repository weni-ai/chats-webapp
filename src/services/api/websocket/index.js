import room from './room';
import discussion from './discussion';
import status from './status';
import soundNotification from './soundNotification';

export default {
  room,
  discussion,
  status,
  soundNotification,
};
