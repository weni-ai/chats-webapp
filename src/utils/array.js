export function removeDuplicatedItems(array, key) {
  const itemsChecked = new Set();
  return array.filter((item) => {
    if (itemsChecked.has(item[key])) {
      return false;
    }
    itemsChecked.add(item[key]);
    return true;
  });
}
