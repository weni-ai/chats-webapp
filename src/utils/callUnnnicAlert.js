import { unnnicCallAlert } from '@weni/unnnic-system';

export default (props) => {
  const mobileMainElement = document.querySelector('.home-mobile__main');

  unnnicCallAlert({
    ...props,
    containerRef: mobileMainElement,
  });
};
