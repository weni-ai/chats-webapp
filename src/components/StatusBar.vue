<template>
  <header
    class="status-bar"
    data-testid="status-bar"
  >
    <section
      class="status-bar__selected"
      data-testid="status-bar-selected"
      @click="toggleDropdown"
    >
      <section class="status-bar__content">
        <section
          class="status-bar__icon"
          data-testid="status-bar-icon"
          :class="`status-bar--${selectedStatus.color}`"
        ></section>
        <p
          class="status-bar__label"
          data-testid="selected-status-label"
        >
          {{ selectedStatus.label }}
        </p>
        <section
          v-if="
            project.config?.can_see_timer &&
            !['inactive', 'active'].includes(selectedStatus.value)
          "
          class="status-bar__timer"
          data-testid="status-bar-timer"
        >
          {{ formattedTime }}
        </section>
      </section>
      <section class="status-bar__set-status">
        <p class="status-bar__set-status__label">
          {{ $t('status.set_status') }}
        </p>
        <UnnnicIcon
          data-testid="header-icon-expand"
          size="md"
          :icon="isOpen ? 'expand_less' : 'expand_more'"
          scheme="neutral-darkest"
        />
      </section>
    </section>

    <ul
      v-if="isOpen"
      class="status-bar__list status-bar__list--open"
      data-testid="status-bar-list-open"
    >
      <li
        v-for="status in statuses"
        :key="status.value"
        class="status-bar__item"
        data-testid="status-bar-item"
        @click="selectStatus(status)"
      >
        <section
          class="status-bar__icon"
          data-testid="status-bar-icon-inside"
          :class="`status-bar--${status.color}`"
        ></section>
        <p
          class="status-bar__item-label"
          data-testid="status-bar-item-label"
        >
          {{ status.label }}
        </p>
      </li>
    </ul>
  </header>
</template>

<script setup>
import { ref, computed, onMounted, onUnmounted } from 'vue';
import { useConfig } from '@/store/modules/config';
import { useProfile } from '@/store/modules/profile';
import { intervalToDuration, parseISO } from 'date-fns';
import api from '@/services/api/resources/chats/pauseStatus';
import Profile from '@/services/api/resources/profile';
import unnnic from '@weni/unnnic-system';
import i18n from '@/plugins/i18n';

const statuses = ref([
  { value: 'active', label: 'Online', color: 'green' },
  { value: 'inactive', label: 'Offline', color: 'gray' },
]);

const selectedStatus = ref(
  sessionStorage.getItem('statusAgent') === 'ONLINE'
    ? statuses.value[0]
    : statuses.value[1],
);
const isOpen = ref(false);
const startDate = ref(null);
const elapsedTime = ref(0);
let intervalId = null;
const configStore = useConfig();
const profileStore = useProfile();
const project = computed(() => configStore.project);
const loadingActiveStatus = ref(false);

const fetchCustomStatuses = async () => {
  const response = await api.getCustomStatusTypeList();
  statuses.value = response;
};

const handleGetActiveStatus = async () => {
  const activeStatus = await configStore.getStatus(configStore.project.uuid);
  configStore.$patch({
    status: activeStatus,
  });
};

const updateActiveStatus = async (isActive) => {
  loadingActiveStatus.value = true;

  const {
    data: { connection_status },
  } = await Profile.updateStatus({
    projectUuid: configStore.project.uuid,
    status: isActive ? 'ONLINE' : 'OFFLINE',
  });

  sessionStorage.setItem('statusAgent', connection_status);

  loadingActiveStatus.value = false;
  showStatusAlert(connection_status.toLowerCase());
};

const startTimer = () => {
  intervalId = setInterval(() => {
    if (startDate.value) {
      const duration = intervalToDuration({
        start: parseISO(startDate.value),
        end: new Date(),
      });
      elapsedTime.value = duration;
    }
  }, 1000);
};

const stopTimer = () => {
  if (intervalId) clearInterval(intervalId);
};

const formattedTime = computed(() => {
  if (!elapsedTime.value) return '00:00:00';
  const { hours = 0, minutes = 0, seconds = 0 } = elapsedTime.value;
  return `${String(hours).padStart(2, '0')}:${String(minutes).padStart(2, '0')}:${String(seconds).padStart(2, '0')}`;
});

const selectStatus = async (newStatus) => {
  if (newStatus.value === selectedStatus.value.value) return;

  const isActiveOrInactive = ['active', 'inactive'].includes(
    selectedStatus.value.value,
  );
  const isCustomStatus = !['active', 'inactive'].includes(newStatus.value);

  if (isActiveOrInactive && isCustomStatus) {
    startDate.value = new Date().toISOString();
    await handleCreateCustomStatus(newStatus);
    startTimer();
  } else if (!isActiveOrInactive && isCustomStatus) {
    await handleCloseCustomStatus(selectedStatus.value, false);
    await handleCreateCustomStatus(newStatus);
    startDate.value = new Date().toISOString();
    startTimer();
  } else if (!isActiveOrInactive && !isCustomStatus) {
    await handleCloseCustomStatus(
      selectedStatus.value,
      newStatus.value === 'active',
    );
    stopTimer();
  }

  if (newStatus.value === 'active' || newStatus.value === 'inactive') {
    updateActiveStatus(newStatus.value === 'active');
  }

  selectedStatus.value = newStatus;
  isOpen.value = false;
};

onMounted(async () => {
  await handleGetActiveStatus();
  await fetchCustomStatuses();
  await getActiveCustomStatusAndActiveTimer();
});

onUnmounted(() => {
  stopTimer();
});

const toggleDropdown = () => {
  isOpen.value = !isOpen.value;
};

const getActiveCustomStatusAndActiveTimer = async () => {
  const activeStatus = await api.getActiveCustomStatus();

  if (activeStatus?.status_type && activeStatus.is_active) {
    statuses.value = statuses.value.map((status) => ({
      ...status,
      statusUuid:
        status.value === activeStatus.status_type ? activeStatus.uuid : null,
    }));

    selectedStatus.value =
      statuses.value.find(
        (status) => status.value === activeStatus.status_type,
      ) || selectedStatus.value;

    startDate.value = activeStatus.created_on;
    startTimer();
  }
};

const handleCloseCustomStatus = async (status, isActive) => {
  const closeStatus = (value) =>
    api.closeCustomStatus({
      statusUuid: value,
      endTime: new Date().toISOString().replace('Z', '+00:00'),
      isActive,
    });

  if (!status.statusUuid) {
    const activeStatus = await api.getActiveCustomStatus();
    if (status.value === activeStatus.status_type) {
      statuses.value = statuses.value.map((state) => ({
        ...state,
        statusUuid:
          state.value === activeStatus.status_type ? activeStatus.uuid : null,
      }));

      return closeStatus(activeStatus.uuid);
    }
  }

  return closeStatus(status.statusUuid);
};

const handleCreateCustomStatus = async (status) => {
  const createStatus = (value) =>
    api.createCustomStatus({
      email: profileStore.me.email,
      statusType: value,
    });

  const response = await createStatus(status.value);

  if (!status.statusUuid) {
    statuses.value = statuses.value.map((state) => ({
      ...state,
      statusUuid: state.value === response.status_type ? response.uuid : null,
    }));
  }

  return response;
};

const showStatusAlert = (connectionStatus) => {
  unnnic.unnnicCallAlert({
    props: {
      text: `${i18n.global.t('status_agent')} ${connectionStatus}`,
      icon: 'indicator',
      scheme:
        connectionStatus === 'online'
          ? 'feedback-green'
          : '$unnnic-color-neutral-black',
      closeText: i18n.global.t('close'),
      position: 'bottom-right',
    },
    seconds: 15,
  });
};
</script>

<style lang="scss" scoped>
.status-bar {
  position: relative;
  display: flex;
  padding: $unnnic-spacing-sm;
  justify-content: space-between;
  align-items: center;
  border-bottom: 1px solid $unnnic-color-neutral-soft;
  border-left: 1px solid $unnnic-color-neutral-soft;
  background: $unnnic-color-neutral-white;
  cursor: pointer;

  &__selected {
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: $unnnic-spacing-xs;
  }

  &__content {
    display: flex;
    align-items: center;
    gap: $unnnic-border-radius-md;
  }

  &__icon {
    width: $unnnic-spacing-xs;
    height: $unnnic-spacing-xs;
    border-radius: 50%;
    background-color: $unnnic-color-neutral-cleanest;
  }

  &--green {
    background-color: $unnnic-color-aux-green-300;
  }

  &--gray {
    background-color: $unnnic-color-neutral-cleanest;
  }

  &--brown {
    background-color: $unnnic-color-aux-orange-500;
  }

  &__timer {
    display: flex;
    padding: 0rem $unnnic-border-radius-md;
    justify-content: center;
    align-items: center;
    gap: $unnnic-spacing-xs;
    border-radius: $unnnic-border-radius-pill;
    border: 1px solid $unnnic-color-neutral-clean;
    color: $unnnic-color-neutral-clean;

    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-md;
    font-style: normal;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;
  }

  &__label {
    color: $unnnic-color-neutral-dark;

    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    font-weight: $unnnic-font-weight-bold;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__set-status {
    display: flex;
    align-items: center;
    gap: $unnnic-spacing-nano;
    &__label {
      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-md;
      font-style: normal;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;
    }
  }

  &__list {
    position: absolute;
    width: 100%;
    background: $unnnic-color-background-snow;
    border-top: 1px solid $unnnic-color-neutral-soft;
    list-style: none;
    margin: 0;
    top: 100%;
    left: 0;
    z-index: 9999999;

    &--open {
      display: block;
    }
  }

  &__item {
    padding: $unnnic-spacing-sm;
    display: flex;
    align-items: center;
    border-bottom: 1px solid $unnnic-color-neutral-soft;
    gap: $unnnic-border-radius-md;

    &:last-child {
      border-bottom: none;
    }

    &:hover {
      background: $unnnic-color-neutral-lightest;
    }

    &-label {
      color: $unnnic-color-neutral-dark;

      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-gt;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
    }
  }
}
</style>
