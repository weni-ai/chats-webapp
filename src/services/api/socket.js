import env from '@/utils/env';
import { getToken, getProject } from '@/utils/config';

export class Socket {
  /**
   * @type {WebSocket}
   */
  #ws;

  #handlers = [];

  constructor(url) {
    this.#ws = new WebSocket(url);
    this.#createOnMessageListener();
  }

  #createOnMessageListener() {
    this.#ws.onmessage = (event) => {
      const data = JSON.parse(event.data);
      const { action } = data;
      const content = JSON.parse(data.content);

      this.#handlers.forEach(({ message, callback }) => {
        if (message === action) callback(content);
      });
    };
  }

  on(message, callback) {
    const handler = {
      message,
      callback,
    };
    this.#handlers.push(handler);
  }
}

const token = getToken();
const project = getProject();

export const ws = new Socket(
  `${env('CHATS_WEBSOCKET_URL')}/agent/rooms?Token=${token}&project=${project}`,
);
