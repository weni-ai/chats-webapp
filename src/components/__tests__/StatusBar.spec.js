import {
  beforeAll,
  afterAll,
  describe,
  expect,
  it,
  vi,
  beforeEach,
  afterEach,
} from 'vitest';
import { mount, config, flushPromises } from '@vue/test-utils';
import StatusBar from '@/components/StatusBar.vue';
import i18n from '@/plugins/i18n';
import api from '@/services/api/resources/chats/pauseStatus';
import { createTestingPinia } from '@pinia/testing';
import Profile from '@/services/api/resources/profile';
import { setActivePinia } from 'pinia';
import unnnic from '@weni/unnnic-system';
import { moduleStorage } from '@/utils/storage';

vi.mock('@weni/unnnic-system', () => ({
  default: {
    unnnicCallAlert: vi.fn(),
  },
}));

vi.mock('@/store/modules/config', () => ({
  useConfig: () => ({
    project: {
      uuid: 'test-uuid',
      config: {
        can_see_timer: true,
      },
    },
    status: 'ONLINE',
    getStatus: vi.fn().mockResolvedValue({
      data: { connection_status: 'ONLINE' },
    }),
    $patch: vi.fn(),
  }),
}));

vi.mock('date-fns', async () => {
  const actual = await vi.importActual('date-fns');
  return {
    ...actual,
    format: vi.fn(() => '2024-02-14T10:00:00+00:00'),
    intervalToDuration: vi.fn(() => ({ hours: 1, minutes: 30, seconds: 45 })),
    parseISO: vi.fn(),
  };
});

vi.mock('@/services/api/resources/chats/pauseStatus', () => ({
  default: {
    getCustomStatusTypeList: vi.fn().mockResolvedValue([
      { value: 'active', label: 'Online', color: 'green' },
      { value: 'inactive', label: 'Offline', color: 'gray' },
      { value: 'lunch', label: 'Lunch', color: 'brown' },
    ]),
    getActiveCustomStatus: vi.fn().mockResolvedValue({
      uuid: 'status-123',
      status_type: 'lunch',
      is_active: true,
      created_on: '2024-02-14T10:00:00+00:00',
    }),
    closeCustomStatus: vi.fn().mockResolvedValue({}),
    createCustomStatus: vi.fn().mockResolvedValue({
      uuid: 'new-status-123',
      status_type: 'lunch',
    }),
  },
}));

vi.mock('@/services/api/resources/profile', () => ({
  default: {
    updateStatus: vi.fn().mockResolvedValue({
      data: { connection_status: 'ONLINE' },
    }),
  },
}));

beforeAll(() => {
  config.global.plugins = config.global.plugins.filter(
    (plugin) => plugin !== i18n,
  );
});

afterAll(() => {
  if (!config.global.plugins.includes(i18n)) {
    config.global.plugins.push(i18n);
  }
});

describe('StatusBar', () => {
  let wrapper;
  let pinia;

  beforeEach(() => {
    pinia = createTestingPinia({
      createSpy: vi.fn,
      initialState: {
        config: {
          project: {
            uuid: 'test-uuid',
            config: {
              can_see_timer: true,
            },
          },
          status: 'ONLINE',
          getStatus: vi.fn().mockResolvedValue({
            data: { connection_status: 'ONLINE' },
          }),
        },
        profile: {
          me: {
            email: 'test@example.com',
          },
        },
      },
    });
    setActivePinia(pinia);
    moduleStorage.clear({ useSession: true });
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.clearAllTimers();
    vi.clearAllMocks();
    if (wrapper) {
      wrapper.unmount();
    }
  });

  const createWrapper = () => {
    return mount(StatusBar, {
      global: {
        plugins: [pinia],
        mocks: {
          $t: (key) => key,
        },
        stubs: {
          UnnnicIcon: {
            template:
              '<div class="unnnic-icon" data-testid="unnnic-icon"><slot /></div>',
            props: ['icon', 'scheme'],
          },
        },
      },
    });
  };

  describe('Initial Rendering', () => {
    it('should render the status bar component', () => {
      wrapper = createWrapper();
      expect(wrapper.find('[data-testid="status-bar"]').exists()).toBe(true);
    });

    it('should initialize with Offline status', () => {
      wrapper = createWrapper();
      expect(wrapper.find('[data-testid="selected-status-label"]').text()).toBe(
        'Offline',
      );
    });
  });

  describe('Dropdown Functionality', () => {
    it('should toggle dropdown when clicking on status bar', async () => {
      wrapper = createWrapper();
      expect(
        wrapper.find('[data-testid="status-bar-list-open"]').exists(),
      ).toBe(false);

      await wrapper
        .find('[data-testid="status-bar-selected"]')
        .trigger('click');
      expect(
        wrapper.find('[data-testid="status-bar-list-open"]').exists(),
      ).toBe(true);

      vi.advanceTimersByTime(300);
      await wrapper.vm.$nextTick();

      await wrapper
        .find('[data-testid="status-bar-selected"]')
        .trigger('click');
      expect(
        wrapper.find('[data-testid="status-bar-list-open"]').exists(),
      ).toBe(false);
    });

    it('should prevent immediate outside clicks from closing dropdown', async () => {
      wrapper = createWrapper();

      await wrapper
        .find('[data-testid="status-bar-selected"]')
        .trigger('click');
      expect(wrapper.vm.isOpen).toBe(true);
      expect(wrapper.vm.isToggling).toBe(true);

      const mockEvent = {
        target: {
          closest: () => null,
        },
      };

      wrapper.vm.handleClickOutside(mockEvent);
      await wrapper.vm.$nextTick();

      expect(wrapper.vm.isOpen).toBe(true);

      vi.advanceTimersByTime(300);
      await wrapper.vm.$nextTick();

      wrapper.vm.handleClickOutside(mockEvent);
      await wrapper.vm.$nextTick();

      expect(wrapper.vm.isOpen).toBe(false);
    });

    it('should ignore rapid successive clicks', async () => {
      wrapper = createWrapper();

      await wrapper
        .find('[data-testid="status-bar-selected"]')
        .trigger('click');
      expect(wrapper.vm.isOpen).toBe(true);
      expect(wrapper.vm.isToggling).toBe(true);

      await wrapper
        .find('[data-testid="status-bar-selected"]')
        .trigger('click');
      expect(wrapper.vm.isOpen).toBe(true);

      vi.advanceTimersByTime(300);
      await wrapper.vm.$nextTick();

      expect(wrapper.vm.isToggling).toBe(false);

      await wrapper
        .find('[data-testid="status-bar-selected"]')
        .trigger('click');
      expect(wrapper.vm.isOpen).toBe(false);
    });
  });

  describe('Custom Status Management', () => {
    beforeEach(() => {
      api.getCustomStatusTypeList.mockResolvedValue([
        { value: 'active', label: 'Online', color: 'green' },
        { value: 'inactive', label: 'Offline', color: 'gray' },
        { value: 'lunch', label: 'Lunch', color: 'brown' },
      ]);

      api.getActiveCustomStatus.mockResolvedValue({
        uuid: '123',
        status_type: 'lunch',
        is_active: true,
        created_on: '2024-02-14T10:00:00+00:00',
      });
    });

    it('should fetch and set custom statuses on mount', async () => {
      wrapper = createWrapper();
      await wrapper.vm.$nextTick();

      await wrapper.vm.fetchCustomStatuses();
      wrapper.vm.toggleDropdown();
      await wrapper.vm.$nextTick();

      const items = wrapper.findAll('[data-testid="status-bar-item"]');
      expect(items).toHaveLength(3);
      expect(wrapper.vm.statuses).toEqual([
        { value: 'active', label: 'Online', color: 'green', statusUuid: null },
        {
          value: 'inactive',
          label: 'Offline',
          color: 'gray',
          statusUuid: null,
        },
        { value: 'lunch', label: 'Lunch', color: 'brown', statusUuid: '123' },
      ]);
      expect(api.getCustomStatusTypeList).toHaveBeenCalled();
    });

    it('should handle switching from active to custom status', async () => {
      wrapper = createWrapper();

      await wrapper.vm.fetchCustomStatuses();
      wrapper.vm.toggleDropdown();
      await wrapper.vm.$nextTick();

      const items = wrapper.findAll('[data-testid="status-bar-item"]');
      await items[2].trigger('click');

      expect(api.createCustomStatus).toHaveBeenCalledWith({
        email: 'test@example.com',
        statusType: 'lunch',
      });
    });

    it('should handle switching between custom statuses', async () => {
      api.getCustomStatusTypeList.mockResolvedValue([
        { value: 'active', label: 'Online', color: 'green' },
        { value: 'inactive', label: 'Offline', color: 'gray' },
        { value: 'lunch', label: 'Lunch', color: 'brown' },
        { value: 'other', label: 'other', color: 'brown' },
      ]);

      api.getActiveCustomStatus.mockResolvedValue({
        uuid: 'status-123',
        status_type: 'lunch',
        is_active: true,
        created_on: '2024-02-14T10:00:00+00:00',
      });

      api.closeCustomStatus.mockClear();
      api.createCustomStatus.mockClear();

      wrapper = createWrapper();
      await flushPromises();

      wrapper.vm.selectedStatus = {
        value: 'lunch',
        label: 'Lunch',
        color: 'brown',
        statusUuid: 'status-123',
      };

      await wrapper.vm.fetchCustomStatuses();
      wrapper.vm.toggleDropdown();
      await wrapper.vm.$nextTick();

      const items = wrapper.findAll('[data-testid="status-bar-item"]');

      await items[3].trigger('click');
      await flushPromises();

      expect(api.closeCustomStatus).toHaveBeenCalledWith(
        expect.objectContaining({
          statusUuid: 'status-123',
          isActive: false,
        }),
      );

      expect(api.createCustomStatus).toHaveBeenCalledWith(
        expect.objectContaining({
          email: expect.any(String),
          statusType: expect.any(String),
        }),
      );
    });

    it('should close custom status when switching to active/inactive', async () => {
      wrapper = createWrapper();
      await wrapper.vm.$nextTick();

      wrapper.vm.selectedStatus = {
        value: 'lunch',
        label: 'Lunch',
        color: 'brown',
        statusUuid: 'status-123',
      };
      wrapper.vm.isOpen = true;
      await wrapper.vm.$nextTick();

      const items = wrapper.findAll('[data-testid="status-bar-item"]');
      await items[0].trigger('click');

      expect(api.closeCustomStatus).toHaveBeenCalled();
    });
  });

  describe('Timer Management', () => {
    it('should start timer for custom status', async () => {
      wrapper = createWrapper();

      await wrapper.vm.fetchCustomStatuses();
      wrapper.vm.toggleDropdown();
      await wrapper.vm.$nextTick();

      const items = wrapper.findAll('[data-testid="status-bar-item"]');
      await items[2].trigger('click');

      expect(wrapper.vm.startDate).not.toBeNull();
      expect(wrapper.find('[data-testid="status-bar-timer"]').exists()).toBe(
        true,
      );
    });

    it('should not show timer for active/inactive status', async () => {
      wrapper = createWrapper();
      await wrapper.vm.$nextTick();

      wrapper.vm.selectedStatus = {
        value: 'active',
        label: 'Online',
        color: 'green',
      };
      await wrapper.vm.$nextTick();

      expect(wrapper.find('[data-testid="status-bar-timer"]').exists()).toBe(
        false,
      );
    });

    it('should format time correctly', () => {
      wrapper = createWrapper();
      wrapper.vm.elapsedTime = { hours: 1, minutes: 30, seconds: 45 };

      expect(wrapper.vm.formattedTime).toBe('01:30:45');
    });
  });

  describe('Status Alerts', () => {
    it('should show success alert when changing to online status', async () => {
      wrapper = createWrapper();
      await wrapper.vm.$nextTick();

      const onlineStatus = { value: 'active', label: 'Online', color: 'green' };
      wrapper.vm.showStatusAlert(onlineStatus, true);

      expect(unnnic.unnnicCallAlert).toHaveBeenCalledWith({
        props: {
          text: 'Status updated to Online',
          icon: 'indicator',
          scheme: 'feedback-green',
          closeText: 'Close',
          position: 'bottom-right',
        },
        seconds: 15,
      });
    });

    it('should show success alert when changing to offline status', async () => {
      wrapper = createWrapper();
      await wrapper.vm.$nextTick();

      const offlineStatus = {
        value: 'inactive',
        label: 'Offline',
        color: 'gray',
      };
      wrapper.vm.showStatusAlert(offlineStatus, true);

      expect(unnnic.unnnicCallAlert).toHaveBeenCalledWith({
        props: {
          text: 'Status updated to Offline',
          icon: 'indicator',
          scheme: '$unnnic-color-neutral-black',
          closeText: 'Close',
          position: 'bottom-right',
        },
        seconds: 15,
      });
    });

    it('should show success alert when changing to custom status', async () => {
      wrapper = createWrapper();
      await wrapper.vm.$nextTick();

      const customStatus = { value: 'lunch', label: 'Lunch', color: 'brown' };
      wrapper.vm.showStatusAlert(customStatus, true);

      expect(unnnic.unnnicCallAlert).toHaveBeenCalledWith({
        props: {
          text: 'Status updated to Lunch',
          icon: 'indicator',
          scheme: 'feedback-green',
          closeText: 'Close',
          position: 'bottom-right',
        },
        seconds: 15,
      });
    });

    it('should show error alert when status update fails', async () => {
      wrapper = createWrapper();
      await wrapper.vm.$nextTick();

      const status = { value: 'active', label: 'Online', color: 'green' };
      wrapper.vm.showStatusAlert(status, false);

      expect(unnnic.unnnicCallAlert).toHaveBeenCalledWith({
        props: {
          text: 'Unable to update status, please try again.',
          icon: 'indicator',
          scheme: 'feedback-red',
          closeText: 'Close',
          position: 'bottom-right',
        },
        seconds: 15,
      });
    });

    it('should show error alert when API request fails', async () => {
      Profile.updateStatus.mockRejectedValueOnce(new Error('API Error'));

      wrapper = createWrapper();
      await wrapper.vm.$nextTick();

      wrapper.vm.isOpen = true;
      await wrapper.vm.$nextTick();

      const items = wrapper.findAll('[data-testid="status-bar-item"]');
      await items[1].trigger('click');
      await flushPromises();

      expect(unnnic.unnnicCallAlert).toHaveBeenCalledWith({
        props: {
          text: 'Unable to update status, please try again.',
          icon: 'indicator',
          scheme: 'feedback-red',
          closeText: 'Close',
          position: 'bottom-right',
        },
        seconds: 15,
      });
    });

    it('should show error alert when custom status creation fails', async () => {
      api.createCustomStatus.mockRejectedValueOnce(new Error('API Error'));

      wrapper = createWrapper();
      await wrapper.vm.fetchCustomStatuses();
      wrapper.vm.toggleDropdown();
      await wrapper.vm.$nextTick();

      const items = wrapper.findAll('[data-testid="status-bar-item"]');
      await items[2].trigger('click');
      await flushPromises();

      expect(unnnic.unnnicCallAlert).toHaveBeenCalledWith({
        props: {
          text: 'Unable to update status, please try again.',
          icon: 'indicator',
          scheme: 'feedback-red',
          closeText: 'Close',
          position: 'bottom-right',
        },
        seconds: 15,
      });
    });
  });

  describe('API Interactions', () => {
    it('should call updateStatus API when changing to active/inactive', async () => {
      wrapper = createWrapper();
      await wrapper.vm.fetchCustomStatuses();
      wrapper.vm.toggleDropdown();
      await wrapper.vm.$nextTick();

      const items = wrapper.findAll('[data-testid="status-bar-item"]');
      await items[1].trigger('click');

      expect(Profile.updateStatus).toHaveBeenCalledWith({
        projectUuid: 'test-uuid',
        status: 'OFFLINE',
      });
    });
  });

  describe('Timer Functionality', () => {
    it('should stop timer when unmounting component', async () => {
      wrapper = createWrapper();
      await wrapper.vm.fetchCustomStatuses();
      wrapper.vm.toggleDropdown();
      await wrapper.vm.$nextTick();

      const items = wrapper.findAll('[data-testid="status-bar-item"]');
      await items[2].trigger('click');

      const clearIntervalSpy = vi.spyOn(window, 'clearInterval');
      wrapper.unmount();

      expect(clearIntervalSpy).toHaveBeenCalled();
    });

    it('should update elapsed time every second', async () => {
      wrapper = createWrapper();
      wrapper.vm.selectedStatus = {
        value: 'lunch',
        label: 'Lunch',
        color: 'brown',
      };
      wrapper.vm.startDate = new Date().toISOString();
      await wrapper.vm.$nextTick();

      vi.advanceTimersByTime(1000);
      await wrapper.vm.$nextTick();

      expect(wrapper.vm.elapsedTime).toBeDefined();
    });
  });

  describe('Session Storage Management', () => {
    it('should initialize status from session storage', async () => {
      moduleStorage.setItem('statusAgent-test-uuid', 'ONLINE', {
        useSession: true,
      });
      wrapper = createWrapper();
      await wrapper.vm.$nextTick();

      expect(wrapper.vm.selectedStatus.value).toBe('active');
    });

    it('should update session storage when changing status', async () => {
      wrapper = createWrapper();
      wrapper.vm.isOpen = true;
      await wrapper.vm.$nextTick();

      const items = wrapper.findAll('[data-testid="status-bar-item"]');
      await items[0].trigger('click');

      expect(
        moduleStorage.getItem('statusAgent-test-uuid', null, {
          useSession: true,
        }),
      ).toBe('ONLINE');
    });
  });

  describe('Status Color Management', () => {
    it('should apply correct color class for each status', async () => {
      wrapper = createWrapper();
      await wrapper.vm.$nextTick();

      const statusIcon = wrapper.find('[data-testid="status-bar-icon"]');
      expect(statusIcon.classes()).toContain('status-bar--green');

      wrapper.vm.selectedStatus = {
        value: 'inactive',
        label: 'Offline',
        color: 'gray',
      };
      await wrapper.vm.$nextTick();
      expect(statusIcon.classes()).toContain('status-bar--gray');

      wrapper.vm.selectedStatus = {
        value: 'lunch',
        label: 'Lunch',
        color: 'brown',
      };
      await wrapper.vm.$nextTick();
      expect(statusIcon.classes()).toContain('status-bar--brown');
    });
  });

  describe('LocalStorage Settings Update Handling', () => {
    beforeEach(() => {
      vi.useFakeTimers();
      Storage.prototype.getItem = vi.fn();
      Storage.prototype.setItem = vi.fn();
    });

    afterEach(() => {
      vi.useRealTimers();
      vi.restoreAllMocks();
    });

    it('should not refresh data when localStorage settingsUpdated remains the same', async () => {
      const getItemMock = vi.fn().mockImplementation((key) => {
        if (key === 'settingsUpdated') return '1234567890';
        if (key === 'lastSettingsUpdate') return '1234567890';
        return null;
      });

      Storage.prototype.getItem = getItemMock;

      wrapper = createWrapper();
      await flushPromises();

      const refreshDataSpy = vi.spyOn(wrapper.vm, 'refreshData');
      refreshDataSpy.mockClear();

      vi.advanceTimersByTime(1000);
      await flushPromises();

      expect(refreshDataSpy).not.toHaveBeenCalled();
    });
  });
});
