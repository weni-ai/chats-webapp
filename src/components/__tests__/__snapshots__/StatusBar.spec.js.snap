// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatusBar > Snapshot > should match snapshot 1`] = `
<header
  class="status-bar"
  data-testid="status-bar"
  data-v-c3dfd30e=""
>
  <section
    class="status-bar__selected"
    data-testid="status-bar-selected"
    data-v-c3dfd30e=""
  >
    <section
      class="status-bar__content"
      data-v-c3dfd30e=""
    >
      <section
        class="status-bar__icon"
        data-testid="status-bar-icon"
        data-v-c3dfd30e=""
        style="background-color: gray;"
      />
      <p
        class="status-bar__label"
        data-testid="selected-status-label"
        data-v-c3dfd30e=""
      >
        Offline
      </p>
      <!--v-if-->
    </section>
    <section
      class="status-bar__set-status"
      data-v-c3dfd30e=""
    >
      <p
        class="status-bar__set-status__label"
        data-v-c3dfd30e=""
      >
        status.set_status
      </p>
      <span
        class="material-symbols-rounded unnnic-icon-scheme--neutral-darkest unnnic-icon-size--md"
        data-testid="header-icon-expand"
        data-v-76ceda5d=""
        data-v-c3dfd30e=""
        translate="no"
      >
        expand_more
      </span>
    </section>
  </section>
  <!--v-if-->
</header>
`;
